Update documentation guidelines and use intersphinx links where possible

laymonage opened the pull request:
Update our documentation guidelines to add more examples and explanation on how to use intersphinx and how to find the targets.

Also update hardcoded links in our docs to use intersphinx, so that we can detect broken links in CI.

Remaining external docs links:
- `django-filter`: There are currently 4 links to their docs. I'd love to add an intersphinx mapping for this, especially if we end up adding more links to their docs in the future. However, currently the sections we link to (e.g. https://django-filter.readthedocs.io/en/stable/guide/usage.html#generating-filters-with-meta-fields) don't have a Sphinx target created. In general, section targets in their docs are quite sparse, as they seem to only add targets as needed. If we add more django-filter links in the future, I'll try sending a PR to them to add some explicit Sphinx targets.
- `django-taggit`: There are 3 links and all of them are to the docs homepage. I think it's not worth adding an intersphinx mapping for now. If we have more links to specific targets, we can add it in the future.
- `pillow`: Three links, all of them point to https://pillow.readthedocs.io/en/stable/installation.html#external-libraries. That link is outdated, and is now found in https://pillow.readthedocs.io/en/stable/installation/building-from-source.html#external-libraries. Not sure if this is worth adding an intersphinx mapping for...
- `elasticsearch-py`: Two links:
  - https://elasticsearch-py.readthedocs.io/en/5.0.0/api.html#elasticsearch
  - https://elasticsearch-py.readthedocs.io/
  Both still work, not worth for intersphinx IMO.
- Django release notes links: there are no explicit targets for sections in their release notes, so these need to be full URLs.
- The rest: random, one-off links:
  - https://github.com/wagtail/wagtail/blob/21206dc3e70f1ecbf618d0d2c80d33f5217f8fa2/docs/deployment/under_the_hood.md?plain=1#L33
  - https://github.com/wagtail/wagtail/blob/21206dc3e70f1ecbf618d0d2c80d33f5217f8fa2/docs/contributing/first_contribution_guide.md?plain=1#L367
  - https://github.com/wagtail/wagtail/blob/21206dc3e70f1ecbf618d0d2c80d33f5217f8fa2/docs/advanced_topics/customization/page_editing_interface.md?plain=1#L184

laymonage commented on docs/conf.py:
```diff
@@ -361,3 +377,26 @@
 def setup(app):
     app.add_js_file("js/banner.js")
+
+    # Django-specific roles, from
+    # https://github.com/django/django/blob/main/docs/_ext/djangodocs.py:
+    app.add_crossref_type(
+        directivename="setting",
+        rolename="setting",
+        indextemplate="pair: %s; setting",
+    )
+    app.add_crossref_type(
+        directivename="templatetag",
+        rolename="ttag",
+        indextemplate="pair: %s; template tag",
+    )
+    app.add_crossref_type(
+        directivename="templatefilter",
+        rolename="tfilter",
+        indextemplate="pair: %s; template filter",
+    )
+    app.add_crossref_type(
+        directivename="fieldlookup",
+        rolename="lookup",
+        indextemplate="pair: %s; field lookup type",
+    )
```
When using the `inv:` syntax in MyST, we don't need to add these to link to the roles. For example, we can already do `[](inv:django:std:setting#DATE_INPUT_FORMATS)`.

However, there's no such syntax in rST. I tried different ways but it seems that Sphinx won't pick up these custom roles. See: https://stackoverflow.com/questions/13387125

Adding this allows us to do `` :setting:`django:DATE_INPUT_FORMATS` `` in rST docs (docstrings and autodoc). It also allows us to use it in MyST like `` {setting}`django:DATE_INPUT_FORMATS` `` (or simply `` {setting}`DATE_INPUT_FORMATS` ``).

laymonage commented on docs/releases/2.16.md:
```diff
@@ -116,7 +116,7 @@ This release upgrades the [django-taggit](https://django-taggit.readthedocs.io/e
 ### `wagtail.admin.views.generic.DeleteView` follows Django 4.0 conventions
 
-The internal (undocumented) class-based view `wagtail.admin.views.generic.DeleteView` has been updated to align with [Django 4.0's `DeleteView` implementation](https://docs.djangoproject.com/en/4.0/releases/4.0/#deleteview-changes), which uses `FormMixin` to handle POST requests. Any custom deletion logic in `delete()` handlers should be moved to `form_valid()`.
+The internal (undocumented) class-based view `wagtail.admin.views.generic.DeleteView` has been updated to align with [Django 4.0's `DeleteView` implementation](https://docs.djangoproject.com/en/stable/releases/4.0/#deleteview-changes), which uses `FormMixin` to handle POST requests. Any custom deletion logic in `delete()` handlers should be moved to `form_valid()`.
```
Django's release notes tend to not have specific Sphinx targets for their sections, so we need to use full URLs. I'm just updating the link to use `stable` instead of the 4.0 version (but still the correct 4.0 notes)

laymonage commented on docs/contributing/documentation_guidelines.md:
```diff
@@ -177,13 +213,13 @@ A [link to a reference](register_reports_menu_item).
 </details>
 
-#### Reference links
+#### Anchor links
 
-Reference links (links to a target within a page) rely on the page having a reference created. Each reference must have a unique name and should use the `lower_snake_case` format. A reference can be added as follows:
```
I changed the terminology from "reference" to "anchor" here because it's a bit ambiguous with "reference" as in "code/API reference"

laymonage commented on docs/contributing/documentation_guidelines.md:
````diff
@@ -211,35 +247,177 @@ Rendered output:
+If the `name` contains a space, you need to wrap the whole link in angle brackets `<>`.
+
+```md
+See Django's docs on [](<inv:django:std:label#topics/cache:template fragment caching>).
````
I didn't know that this was possible, so TIL

https://github.com/wagtail/wagtail/pull/9603#discussion_r1246402586

laymonage commented:
@lb- Hopefully this will make it easier for contributors to use intersphinx links whenever possible in future docs updates

laymonage commented on docs/contributing/documentation_guidelines.md:
```diff
@@ -211,35 +247,177 @@ Rendered output:
+
+</details>
+
+##### Finding intersphinx targets
+
+The intersphinx target for a specific anchor you want to link to may not be obvious. You can use the `myst-inv` command line tool from MyST-Parser and save the output as a JSON or YAML file to get a visual representation of the available targets.
```
For me personally I think this was the most useful section of the entire page 😆

lb- replied:
Super great. I wonder if there's a website that has them all loaded for exploration also?

laymonage replied:
I don' t think there is one. Though I think you just gave me an idea for a side project 👀

lb- commented on docs/contributing/documentation_guidelines.md:
``````diff
@@ -157,17 +169,41 @@ some command
+print("Hello, world!")
+```
+````
+`````
``````
Code block inception

lb- commented on docs/contributing/documentation_guidelines.md:
```diff
@@ -157,17 +169,41 @@ some command
 
 Links are fundamental in documentation. Use internal links to tie your content to other docs, and external links as needed. Pick relevant text for links, so readers know where they will land.
 
-Don’t rely on [`links over code`](https://www.example.com/), as they are impossible to spot.
```
Maybe we should refine the nuance here, I always thought this was a good guideline to follow. 

Maybe something like, 

```
Don't let external links hide [critical context](https://www.example.com/) for the reader, instead provide the core information on the page and use links for added context.
```

laymonage replied:
I do not understand what the original sentence meant. It seems to be talking about inline code as links specifically, and I don't find "they are impossible to spot" to be true. Maybe it was a bit harder to spot when the theme used custom fonts, but now with the system fonts combined with the bright underline and the hover effect I think they're not impossible to spot. Otherwise, with this guideline, how would you link to API references?

Your sentence, on the other hand, is much clearer on what makes it not OK. I like it, thanks. I've added your sentence to the section, and added a few examples of anti-patterns.

lb- replied:
Thanks, hopefully that's useful.

lb- commented on docs/getting_started/tutorial.md:
```diff
@@ -199,7 +199,7 @@ to the model. Wagtail uses normal Django templates to render each page
 separating capital letters with underscores. For example, `HomePage` within the "home" app becomes
 `home/home_page.html`. This template file can exist in any location that
-[Django's template rules](https://docs.djangoproject.com/en/stable/intro/tutorial03/#write-views-that-actually-do-something) recognize. Conventionally, you can place it within a `templates` folder within the app.
+[Django's template rules](<inv:django#intro/tutorial03:write views that actually do something>) recognize. Conventionally, you can place it within a `templates` folder within the app.
```
Should this one have the angles brackets?

lb- replied:
Ignore this, it's intentional.

lb- commented on docs/topics/writing_templates.md:
```diff
@@ -331,7 +331,7 @@ if the page supports [multiple preview modes](wagtail.models.Page.preview_modes)
 ## Template fragment caching
 
-Django supports [template fragment caching](https://docs.djangoproject.com/en/stable/topics/cache/#template-fragment-caching), which allows caching portions of a template. Using Django's `{% cache %}` tag natively with Wagtail can be [dangerous](https://github.com/wagtail/wagtail/issues/5074) as it can result in preview content being shown to end users. Instead, Wagtail provides 2 extra template  ...
+Django supports [template fragment caching](<inv:django:std:label#topics/cache:template fragment caching>), which allows caching portions of a template. Using Django's `{% cache %}` tag natively with Wagtail can be [dangerous](https://github.com/wagtail/wagtail/issues/5074) as it can result in preview content being shown to end users. Instead, Wagtail provides 2 extra template tags which can be l ...
```
Same here, how come these use the angled bracki?

lb- replied:
Ahh. I see, because it has spaces. All good.

laymonage commented:
Thanks @lb-. I changed the "Finding intersphinx targets" heading to "Find the right intersphinx target" per Google's styleguide, and added your guideline about not hiding critical context behind links.

lb- reviewed (approved):
Thanks @laymonage  - this is an awesome win.

Happy to approve, I will not get a chance to merge tonight but will do tomorrow if you have not already merged.

laymonage commented on docs/contributing/documentation_guidelines.md:
````diff
@@ -289,27 +475,63 @@ Images are hard to keep up-to-date as documentation evolves, but can be worthwhi
-With its [autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) feature, Sphinx supports writing documentation in Python docstrings for subsequent integration in the project’s documentation pages. This is a very powerful feature that we highly recommend using to document Wagtail’s APIs.
+With its [autodoc](inv:sphinx#ext-autodoc) feature, Sphinx supports writing documentation in Python docstrings for subsequent integration in the project’s documentation pages. This is a very powerful feature that we highly recommend using to document Wagtail’s APIs.
+
+Modules, classes, and functions can be documented with docstrings. Class and instance attributes can be documented with docstrings (with triple quotes `"""`) or doc comments (with hash-colon `#:`). Docstrings are preferred, as they have better integration with code editors. Docstrings in Python code must be written in reStructuredText syntax.
+
+```py
+SLUG_REGEX = r'^[-a-zA-Z0-9_]+$'
+"""Docstring for module-level variable ``SLUG_REGEX``."""
+
+class Foo:
+    """Docstring for class ``Foo``."""
+
+    bar = 1
+    """Docstring for class attribute ``Foo.bar``."""
+
+    #: Doc comment for class attribute ``Foo.baz``.
+    #: It can have multiple lines, and each line must start with ``#:``.
+    #: Note that it is written before the attribute.
+    #: While Sphinx supports this, it is not recommended.
+    baz = 2
````
I didn't know you can write docstrings for attributes too, I thought you had to use the hash-colon syntax.

Apparently docstrings are better supported in code editors. See the following:

![image](https://github.com/user-attachments/assets/cbd444be-9253-45f2-9ff1-ecc7ec617759)


![image](https://github.com/user-attachments/assets/466dcf16-5109-455b-9731-782d71222b5b)

While I don't think we've adopted this pattern for attributes, I think we should use it moving forward. It's a big win for editor integration.

lb- replied:
hmm, nice.

laymonage reviewed (commented):
@lb- I added two commits:
- Remove e.g. in the doc guidelines (as that is against our guidelines 😆): [`a81cc73` (#12525)](https://github.com/wagtail/wagtail/pull/12525/commits/a81cc7345d0348e49787147be4efb04adb8d3b12)
- Expand section on Autodoc to be more about writing docstrings and API reference [`794fa3b` (#12525)](https://github.com/wagtail/wagtail/pull/12525/commits/794fa3bcaf6d2403b369c0c4852ba85469400ffc)

Will merge after lunch, hope that's OK. Thanks for the review!

lb- commented on docs/contributing/documentation_guidelines.md:
```diff
@@ -289,27 +475,63 @@ Images are hard to keep up-to-date as documentation evolves, but can be worthwhi
 
 </details>
 
-### Autodoc
+### Docstrings and API reference (autodoc)
```
Great.
